### ISSUE TITLE
[networkd] Could not set route without gateway (but works on reload)

### ISSUE THREAD

**Issue by Malvineous** [systemd/systemd#19285]
**systemd version the issue has been seen with**
 > v248

**Used distribution**
 > Arch Linux

**Linux kernel version used** (`uname -a`)
 > `5.11.12-arch1-1`

**CPU architecture issue was seen on**
 > x86-64

**Expected behaviour you didn't see**
 > I expected the system to boot normally.

**Unexpected behaviour you saw**
 > System dropped into single user mode, due to failed network mounts.  Running `systemctl restart systemd-networkd` then exiting single-user mode fixes the problem, but this command has to be run in single-user mode on every boot.

**Steps to reproduce the problem**
 > Reboot PC.  The problem cannot be reproduced once the PC has booted as `systemd-networkd` always reuses the existing state and never starts from scratch as it does during the initial boot process.

**Additional program output to the terminal or log subsystem illustrating the issue**

I have a point-to-point link set up, where a cable directly connects two machines via a dedicated network interface (here called `bond10g`).  Both machines have another interface they use for normal traffic which goes to a switch (`eno3`, not shown here).

When booting the system, the interface initially fails to be configured:
```
systemd-networkd[1447]: bond10g: Could not set route without gateway: Invalid prefsrc address. Invalid argument
systemd-networkd[1447]: bond10g: Failed
```
However immediately restarting `systemd-networkd` results in successful configuration, so it is difficult to troubleshoot as it requires a full reboot (~10 minutes on this server) to get back to a clean state.

The configuration for the bond10g interface is:

```
[Match]
Name=bond10g

[Network]
DefaultRouteOnDevice=false

[Route]
Scope=link
PreferredSource=1.1.1.1
Destination=1.1.1.2/32
GatewayOnLink=yes  # Tried adding this but it didn't help, see below
```
If I add `GatewayOnLink=yes` then I get a different error in the logs:
```
bond10g: Could not set route without gateway: Invalid flags for nexthop - PERVASIVE and ONLINK can not be set. Invalid argument
```
If I add `Gateway=1.1.1.2` (which is wrong as there's no gateway on this peer-to-peer link) then I get an error saying there was no route to the host.

I think the problem is that the `bond10g` interface gets configured first, and for some reason it can't be set up without a gateway (even though it's a peer-to-peer link so no gateway is needed.)  However once the machine's main interface is configured (with a gateway) then `systemd-networkd` is able to configure `bond10g` without any issues.

Is there a way to either:

1. Tell `systemd-networkd` to configure one interface before another?  I tried renaming the files so they had different numbers at the start but this didn't make any difference.
2. Run `systemctl restart systemd-networkd` during boot so that it reloads itself during the boot process, fixing the issue without me having to run that command manually in the middle of each boot?
3. A config option I have missed.  I've gone through the manpages and tried a bunch of things but none of them made any difference except the ones mentioned above.

This was working with earlier versions of systemd but it's been about six months since I last upgraded so no doubt the change that broke my config was introduced quite some time ago.  I note there is now a new multipath option in the manpage which wasn't there before, so I wonder whether that perhaps interfered with this type of config as it seems it requires a gateway in its parameters.  I'm only guessing though.

**Comment by yuwata** [comment]
So, another interface (`eno3`?) has the ip address `1.1.1.1`? Could you show all .network files?

**Comment by Malvineous** [comment]
Sorry yes, here is the config for `eno3`.  Nothing unusual here, just a static config.
```
[Match]
Name=eno3

[Network]
Address=1.1.1.1/16
DNS=1.1.1.2
DNS=1.1.1.1
DHCP=no

[Route]
Gateway=1.1.1.3
```
The only other ones probably aren't relevant, they're the bonding config for `bond10g`:
```
# 40-bond10g.network
[Match]
Name=eno1
Name=eno2

[Network]
Bond=bond10g

# bond10g.netdev
[NetDev]
Name=bond10g
Description=2x10G link to 1.1.1.2
Kind=bond

[Bond]
Mode=balance-rr
```

**Comment by yuwata** [comment]
Fix is waiting in #19445.
